File: echo_server/echo_server2.zig
const std = @import("std");
const print = std.debug.print;
const os = std.os;
const net = std.net;
const port = 8000;

pub fn main() !void{
    const server = try os.socket(os.AF.INET, os.SOCK.STREAM, 0);
    defer os.closeSocket(server);

    const addr = try net.Address.resolveIp("127.0.0.1", port);

    try os.bind(server, &addr.any, addr.getOsSockLen());
    try os.listen(server, 1);
    print("listening on port {}\n", .{port});

    const client = try os.accept(server, null, null, 0);
    defer os.closeSocket(client);
    print("client connected!\n", .{});

    const welcome = "Welcome to Zig echo2!\n";
    _= try os.send(client, welcome[0..], 0);

    while ( true ) {
        var buf = [_]u8{0} ** 128;

        const len = try os.recv(client, buf[0..], 0);

        var reply_buf = [_]u8{0} ** 128;
        var fba = std.heap.FixedBufferAllocator.init(&reply_buf).allocator();
        var reply = std.ArrayList(u8).init(fba);

        if ( len > 0 ) {
            if (std.mem.indexOf(u8, buf[0..], "bye")) |_| {
                print("saying goodbye!", .{});
                _=try reply.writer().write("Good bye!");
                _=try os.send(client, reply.items, 0);
                break;
            }
            print("received: {s}", .{buf[0..len]});
            _=try reply.writer().write("echo: ");
            _=try reply.writer().write(buf[0..len]);
            _=try reply.writer().write("\n");
            _=try os.send(client, reply.items, 0);
        } else break;
    }
    print("shutting down\n", .{});
}

File: echo_server/echo_server2.cc
#include <asio.hpp>
#include <stdio.h>
#include <system_error>
#include <string>

int main() {
    asio::io_context io;
    asio::ip::tcp::acceptor server(io, asio::ip::tcp::endpoint(asio::ip::tcp::v4(), 8000));
    puts("Listening on port 8000");

    asio::ip::tcp::socket client(io);
    server.accept(client);
    puts("client connected!");

    std::error_code e;
    asio::write(client, asio::buffer("Welcome to Asio Chat!\n"), e);

    if (e) { printf("greet error %s\n", e.message().c_str()); return -1; }

    while (true) {
        asio::streambuf buf;
        asio::read_until(client, buf, "\n", e);

        if (e == asio::error::eof) { 
            puts("client disconnected");
            break;
        } else if (e) {
            printf("read error %s\n", e.message().c_str()); 
            return -1; 
        }

        std::string msg = asio::buffer_cast<const char*>(buf.data());
        if (msg.find("bye") != std::string::npos) {
            puts("its so hard to say goodbye!");
            asio::write(client, asio::buffer("good bye!\n"), e);
            client.shutdown(asio::ip::tcp::socket::shutdown_both, e);
            client.close(e);
            break;
        }

        auto reply = "echo: " + msg;
        asio::write(client, asio::buffer(reply), e);

        if (e) { printf("write error %s\n", e.message().c_str()); return -1; }

        printf("%s", msg.data());
    }

    puts("shutting down");
}